### PR TITLE
286/test hbgdataset

### PR DESCRIPTION
**Pull request by gnayuy**
In current status, we fix ephys_nwb_data loading nwb file issues, change class name with HBGNWB2Data, and some abstract functions.

I followed Sergey's approach to create the test. The test file test_ephys_nwb_data.py stays in his branch.

### CODE REVIEW

File: ipfx/dataset/ephys_nwb_data.py [hunk 1]
```diff
@@ -63,13 +64,17 @@ def __init__(self,
...
+        
         if load_into_memory:
             with open(nwb_file, 'rb') as fh:
-                nwb_file = BytesIO(fh.read())
+                data = BytesIO(fh.read())
 
-        self.nwb_file = nwb_file
-        self.nwb = NWBHDF5IO(nwb_file, mode='r').read()
+            _h5_file = h5py.File(data, "r+")
+            self.nwb = NWBHDF5IO(path=_h5_file.filename, mode="r+",file=_h5_file).read()
```
Comment (NileGraddis) [inline ipfx/dataset/ephys_nwb_data.py]: mode should be r

File: tests/test_hbg_dataset.py [hunk 2]
```diff
@@ -62,24 +64,24 @@ def test_main_dat(ontology, NWB_file):
...
-@pytest.mark.parametrize('ontology, NWB_file', [(None, 'H18.28.015.11.14.nwb')], indirect=True)
+@pytest.mark.parametrize('ontology, NWB_file', [(None, 'H20.28.008.11.05-10.nwb')], indirect=True)
 def test_get_clamp_mode(ontology, NWB_file):
 
-    dataset = HBGDataSet(nwb_file=NWB_file, ontology=ontology)
+    dataset = HBGNWB2Data(nwb_file=NWB_file, ontology=ontology)
     assert dataset.get_clamp_mode(10101) == dataset.VOLTAGE_CLAMP
 
 
-@pytest.mark.parametrize('ontology, NWB_file', [(None, 'H18.28.015.11.14.nwb')], indirect=True)
```
Comment (sgratiy) [inline tests/test_hbg_dataset.py]: why not use the original file? it should be nwb version 2.

File: tests/test_hbg_dataset.py [hunk 3]
```diff
@@ -22,29 +24,29 @@ def ontology():
                              ])
 
 
-@pytest.mark.parametrize('ontology, NWB_file', [(None, '2018_03_20_0005.nwb')], indirect=True)
-def test_main_abf(ontology, NWB_file):
+# @pytest.mark.parametrize('ontology, NWB_file', [(None, '2018_03_20_0005.nwb')], indirect=True)
```
Comment (sgratiy) [inline tests/test_hbg_dataset.py]: should probably remove commented out lines

File: ipfx/dataset/ephys_nwb_data.py [hunk 4]
```diff
@@ -213,6 +218,33 @@ def get_stimulus_unit(self, sweep_number):
 
             return unit_str
 
+    def get_sweep_record(self, sweep_number: int) -> Dict[str,Any]:
```
Comment (sgratiy) [inline ipfx/dataset/ephys_nwb_data.py]: this method is not needed

File: tests/test_hbg_dataset.py (hunk 2 above)
Comment (gnayuy) [reply tests/test_hbg_dataset.py]: modified the test_hbg_data.py file avoid using the original file

File: tests/test_hbg_dataset.py (hunk 3 above)
Comment (gnayuy) [reply tests/test_hbg_dataset.py]: removed

File: ipfx/dataset/ephys_data_interface.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 from typing import Dict, Any, List, Optional
 import abc
 import warnings
-from ipfx.stimulus import StimulusOntology
+from ipfx.dataset.stimulus import StimulusOntology
```
Comment (NileGraddis) [inline ipfx/dataset/ephys_data_interface.py]: we've moved this back as discussed in slack.

File: ipfx/dataset/ephys_nwb_data.py [hunk 6]
```diff
@@ -63,13 +64,17 @@ def __init__(self,
...
+        self.nwb_file = nwb_file
+        
         if load_into_memory:
             with open(nwb_file, 'rb') as fh:
-                nwb_file = BytesIO(fh.read())
+                data = BytesIO(fh.read())
 
-        self.nwb_file = nwb_file
-        self.nwb = NWBHDF5IO(nwb_file, mode='r').read()
+            _h5_file = h5py.File(data, "r")
```
Comment (NileGraddis) [inline ipfx/dataset/ephys_nwb_data.py]: might want to hang onto this so methods that rely on h5 access don't have to reach through the nwbfile

File: tests/test_hbg_dataset.py [hunk 7]
```diff
@@ -1,11 +1,13 @@
 from __future__ import absolute_import
 import pytest
+import os
```
Comment (NileGraddis) [inline tests/test_hbg_dataset.py]: Why are there two such test files?

File: tests/dataset/test_hbg_dataset.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                             [('name', 'Chirp stimulus'), ('code', 'Chirp')],
+                             [('name', 'extpexpend stimulus'), ('code', 'extpexpend')]
+                             ])
+
+    return HBGNWB2Data(nwb_file=tmp_nwb_path, ontology=ontology)
+
+def test_create_hbg(hbg_nwb_data):
+    assert isinstance(hbg_nwb_data,HBGNWB2Data)
+
+def test_main_dat(hbg_nwb_data):
```
Comment (sgratiy) [inline tests/dataset/test_hbg_dataset.py]: here you are testing a particular function `get_sweep_record` thus the name of the test should be `test_get_sweep_record`. Then you also need a test for get_stim_code_ext. Why are you calling it test_main_dat given that the test is using a fake nwb file?

File: tests/dataset/test_hbg_dataset.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        'leak_pa': 100.0,
+        'stimulus_scale_factor': 32.0,
+        'stimulus_code': 'STIMULUS_CODE',
+        'stimulus_code_ext': 'STIMULUS_CODE',
+        'stimulus_name': 'Unknown'}
+
+    obtained = hbg_nwb_data.get_sweep_record(sweep_num=4)
+    assert expected == obtained
+
+# @pytest.mark.parametrize('ontology, NWB_file', [(None, 'H20.28.008.11.05-10.nwb')], indirect=True)
```
Comment (sgratiy) [inline tests/dataset/test_hbg_dataset.py]: comments should go away.

File: tests/dataset/test_hbg_dataset.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+@pytest.fixture
+def hbg_nwb_data(tmp_nwb_path):
+
+    nwbfile = nwbfile_to_test()
+    print(tmp_nwb_path)
+
+    with pynwb.NWBHDF5IO(path=tmp_nwb_path, mode="w") as writer:
+        writer.write(nwbfile)
+
+    ontology =  StimulusOntology([[('name', 'ramp stimulus'), ('code', 'RAMP1')],
```
Comment (sgratiy) [inline tests/dataset/test_hbg_dataset.py]: For test the ontology should correspond to the test nwb file.  You use "STIMULUS_CODE" but this code is absence from the ontology.

**Comment by gnayuy** [comment]
The unit test for hbg data created for refactored the new interface and also refactored NWB loading func by accepting path and h5 obj.

**Pull request closed without merging**